### ISSUE TITLE
Assertion failed: (Node2Index[SU->NodeNum] > Node2Index[I->getSUnit()->NodeNum] && "Wrong topological sorting"), function InitDAGTopologicalSorting, file /Users/andy/git/llvm-dcpu16/lib/CodeGen/ScheduleDAG.cpp, line 469. 

### ISSUE THREAD

**Issue by a1k0n** [llvm-dcpu16/llvm-dcpu16#190]
I haven't been able to find a simpler test case so far, but I have been able to isolate the crash to a particular store instruction.  The code to repro is at https://github.com/a1k0n/tetris-dcpu16/tree/llvm-crash

I have some pre-compiled llvm assembler files in my llvm-crash branch: crash21.ll which crashes llc, and crash22.ll which doesn't crash llc where I commented out a single store instruction.

It seems that our DAG contains a cycle, but only when a particular store instruction is present.  The culprit in the past for other targets was ISelDAGToDAG but I'm not sure whether that's the case for us.

**Comment by a1k0n** [comment]
Minimal reproducing test case: it's the stores resulting from add-with-carry operations (AKA ADX).

```
extern long x;
extern long y;

void f() {
  y = 1;
  x++;
}
```

The stores get combined into TokenFactors (or something?) but somehow it creates a cycle in the DAG where the glued ADD/ADC chain become both predecessors and successors of the TokenFactor node.  I haven't tracked down yet where these TokenFactors are coming from -- it's not from anything in Target/DCPU16 directly though.  The equivalent 64-bit code works ok on 32-bit x86, so it's not a general llvm problem.

I'm not really sure what's going on; perhaps the guys on IRC can advise.

**Comment by Blei** [comment]
No progress yet, but you (i.e. everyone looking at this report ;) ) can visualize the problem by compiling the above program to an .ll file and then running `bin/llc -view-sched-dags crash.ll`. The problem is the triangle of TokenFactor, ADD16mr and ADD16rm in the middle.

**Comment by Blei** [comment]
So I was looking into this a bit more and asked the irc people about it.

Here are screenshots of the DAG [before instruction selection](http://img802.imageshack.us/img802/8800/iselinputgraph.png) and [after](http://img19.imageshack.us/img19/1333/iseloutputgraph.png). If you look at these two graphs long enough, you notice that nodes 11 and 15 from the input are combined into the ADC16rm in the output, and nodes 12, 13 and 16 are combined into ADD16mr.

The problem is now, that the ADD16mr node points to the TokenFactor (TF) that was between the pair of loads and stores in the input graph (id 14), which in turn points to the ADC16rm node which incorporates the load from the input graph. Also you might notice, that the aforementioned TF isn't actually needed in this case, and that the cycle would go away if it weren't there.

So the guy on IRC (thanks, baldrick) told me: "I suggest you only do the transform [load/add/store -> ADD16mr] when the store uses the chain of the load directly, and spend your effort on removing pointless token factors", "if (1) the store is the only user of chain of the load, (2) the operation is the only user of the loaded value, (3) the store is the only user of the operation result, then it should be fine"

**Comment by Blei** [comment]
Ok, I tried to remove the redundant TF and it seems to work. But only the cases for ADD/SUB and Load and Store are implemented. Not sure how ugly or feasible a complete implementation is.

**Comment by a1k0n** [comment]
Interesting.  The code generating the TokenFactors is LegalizeTypes/LegalizeIntegerTypes, right?  Did you modify that, or add specialized DCPU16 code?  Can you push your WIP branch?

**Comment by Blei** [comment]
Pushed.

**Comment by a1k0n** [comment]
Wow, nice work.  So you separated the hi and lo chains so there's no tokenfactor combining them?

**Comment by Blei** [comment]
Yes, exactly. I'm still generating a TF, but register the individual chains for the hi and lo parts. If there is an operation that needs a chain and is also split, it looks up the registered individual chains and uses those instead of the TF.

**Comment by a1k0n** [comment]
So what are the chances of getting this fixed upstream?  We need a test case (it ought to break spectacularly on MSP430 but I haven't tried) and it looks like its free of the BitsPerByte junk, so it should be possible?  Are you sitting on this because you also want to handle more double-precision int ops besides add/sub?

**Comment by Blei** [comment]
I think I might have gotten all the cases of split chains. There is one problem though: one additional case where a chain is split is here:

``` patch
diff --git a/lib/CodeGen/SelectionDAG/LegalizeTypesGeneric.cpp b/lib/CodeGen/SelectionDAG/LegalizeTyp
index 51ff515..1dbe76e 100644
--- a/lib/CodeGen/SelectionDAG/LegalizeTypesGeneric.cpp
+++ b/lib/CodeGen/SelectionDAG/LegalizeTypesGeneric.cpp
@@ -423,7 +423,9 @@ SDValue DAGTypeLegalizer::ExpandOp_NormalStore(SDNode *N, unsigned OpNo) {
                     St->getPointerInfo().getWithOffset(IncrementSize),
                     isVolatile, isNonTemporal,
                     MinAlign(Alignment, IncrementSize));
-  return DAG.getNode(ISD::TokenFactor, dl, MVT::Other, Lo, Hi);
+  SDValue newChain = DAG.getNode(ISD::TokenFactor, dl, MVT::Other, Lo, Hi);
+  SetExpandedChain(newChain, Lo, Hi);
+  return newChain;
 }
```

Unfortunately this makes a test fail and I haven't really looked into the reason for that yet. Apparently the same chain has already been registered, which is impossible because it is created just three lines above... Broken hashtable?

And I guess we'd need a test, I was too lazy to find one yet ;)
